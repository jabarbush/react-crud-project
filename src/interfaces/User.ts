interface User {
    id: number;
    name: string;
    dob: string;
    phone: string;
    email: string;
    signature?: {
        pin: string;
        fontStyle: string;
    } | null;
  }
  
  export default User;
  