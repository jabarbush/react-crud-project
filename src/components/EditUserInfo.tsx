import React, { useEffect, useState } from 'react';
import EditUserInfoProps from '../interfaces/EditUserInfoProps';
import { ReactComponent as FormClose } from '../assets/form-close.svg';
import User from '../interfaces/User';
import { validateDateOfBirth, validateEmail, validatePhone, validateRequiredFields } from '../utils/FormValidation';

const EditUserInfo: React.FC<EditUserInfoProps> = ({ selectedUser, onCloseEditUserInfo, setUserList }) => {
  const [name, setName] = useState(selectedUser?.name);
  const [dob, setDob] = useState(selectedUser?.dob);
  const [phone, setPhone] = useState(selectedUser?.phone);
  const [email, setEmail] = useState(selectedUser?.email);
  const [initialRender, setInitialRender] = useState(true);

  useEffect(() => {
    if (initialRender) {
      setInitialRender(false);
    } else {
      onCloseEditUserInfo();
    }
  }, [selectedUser]);

  const handleSaveUserInfo = (e: React.FormEvent) => {
    e.preventDefault();
  
    if (!selectedUser) {
      return;
    }
  
    const validatedName = name ?? '';
    const validatedDob = dob ?? '';
    const validatedPhone = phone ?? '';
    const validatedEmail = email ?? '';
  
    if (
      !validateRequiredFields(validatedName, validatedDob, validatedPhone, validatedEmail) ||
      !validateDateOfBirth(validatedDob) ||
      !validatePhone(validatedPhone) ||
      !validateEmail(validatedEmail)
    ) {
      return;
    }
  
    const updatedUser: User = {
      ...selectedUser,
      name: validatedName,
      dob: validatedDob,
      phone: validatedPhone,
      email: validatedEmail,
    };

    fetch(`http://localhost:3001/users/${selectedUser.id}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({...updatedUser,
        signature: selectedUser.signature}),
    })
      .then((response) => {
        if (response.ok) {
          setUserList((prevUserList: User[]) => {
            const updatedUserList = prevUserList.map((user: User) => {
              if (user.id === updatedUser.id) {
                return {...updatedUser,
                  signature: user.signature};
              }
              return user;
            });
            return updatedUserList as User[];
          });
          console.log('User updated successfully');
        } else {
          console.error('Error updating user:', response.statusText);
        }
      })
      .catch((error) => {
        console.error('Error updating user:', error);
      });
  };
  
  return (
    <div className="edit-user-info">
        <h3 className='edit-user-info-title'>Edit User Information</h3>
        <FormClose className='icons icon-big-x' onClick={onCloseEditUserInfo} />
        <br/>
        <div>
            <label>Name</label>
            <input type='text' value={name} onChange={(e) => setName(e.target.value)} required/>
            <label>Date of Birth</label>
            <input type='text' value={dob} onChange={(e) => setDob(e.target.value)} required/>
            <label>Phone</label>
            <input type='text' value={phone} onChange={(e) => setPhone(e.target.value)} required/>
            <label>Email</label>
            <input type='text' value={email} onChange={(e) => setEmail(e.target.value)} required/>
            <div className='signature-btns'>
                  <button className="add-signature-submit" onClick={handleSaveUserInfo} type="submit">Save</button>
                  <button className="cancel-signature" onClick={onCloseEditUserInfo}>Cancel</button>
            </div>
        </div>
    </div>
  );
};

export default EditUserInfo;
