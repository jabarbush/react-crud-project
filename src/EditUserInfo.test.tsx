import { render, screen } from '@testing-library/react';
import User from './interfaces/User';
import EditUserInfo from './components/EditUserInfo';

beforeEach(() => jest.resetAllMocks())
describe('EditUserInfo component', () => {
    const selectedUser = {
        id: 1,
        name: 'jake',
        dob: '03/03/2003',
        phone: '4564564564',
        email: 'jake4@gmail.com',
        signature: {
          pin: '123123',
          fontStyle: 'Mr Dafoe',
        },
      };
    const userList: User[] = [{
        id: 1,
        name: 'jake',
        dob: '03/03/2003',
        phone: '4564564564',
        email: 'jake4@gmail.com',
        signature: {
            pin: '123123',
            fontStyle: 'Mr Dafoe',
        },
    },
    {
        "id": 2,
        "name": "jake425525333",
        "dob": "03/03/2003",
        "phone": "4564564564",
        "email": "jake4@gmail.com",
        "signature": null
      },
      {
        "id": 3,
        "name": "ashley1234",
        "dob": "05/05/1995",
        "phone": "0980980988",
        "email": "ashley@jake.com",
        "signature": {
          "pin": "123123",
          "fontStyle": "Dancing Script"
        }
      },
      {
        "id": 4,
        "name": "123",
        "dob": "01/01/2001",
        "phone": "1231231231",
        "email": "jake@gmail.com",
        "signature": {
          "pin": "123123",
          "fontStyle": "Great Vibes"
        }
      }];

    const setUserList = jest.fn();
    const onCloseEditUserInfo = jest.fn();

    test('EditUserInfo component renders', () => {
        render(
            <EditUserInfo
                selectedUser={selectedUser}
                setUserList={setUserList}
                onCloseEditUserInfo={onCloseEditUserInfo}
            />
        );
        const editTitle = screen.getByText('Edit User Information');
        expect(editTitle).toBeInTheDocument();
    });

    test('EditUserInfo component matches snapshot', () => {
      
        const { asFragment } = render(
            <EditUserInfo
            selectedUser={selectedUser}
            setUserList={setUserList}
            onCloseEditUserInfo={onCloseEditUserInfo}
        />
        );
      
        expect(asFragment()).toMatchSnapshot();
      });
});