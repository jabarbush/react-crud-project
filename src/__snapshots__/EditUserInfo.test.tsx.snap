// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditUserInfo component EditUserInfo component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="edit-user-info"
  >
    <h3
      class="edit-user-info-title"
    >
      Edit User Information
    </h3>
    <svg
      class="icons icon-big-x"
    >
      form-close.svg
    </svg>
    <br />
    <div>
      <label>
        Name
      </label>
      <input
        required=""
        type="text"
        value="jake"
      />
      <label>
        Date of Birth
      </label>
      <input
        required=""
        type="text"
        value="03/03/2003"
      />
      <label>
        Phone
      </label>
      <input
        required=""
        type="text"
        value="4564564564"
      />
      <label>
        Email
      </label>
      <input
        required=""
        type="text"
        value="jake4@gmail.com"
      />
      <div
        class="signature-btns"
      >
        <button
          class="add-signature-submit"
          type="submit"
        >
          Save
        </button>
        <button
          class="cancel-signature"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
